source/doc-pages/guides/basics/repositories.md

```diff
@@ -0,0 +1,309 @@
+# Repositories
+
+<div class="well">
+This Guide is based on rom 0.9.0.beta and rom-repository 0.0.2.
+</div>
+
+In ROM repositories encapsulates access to domain-specific entities. They use
+relations to expose a convenient interface for composing relations together
+and automatically mapping them to struct objects.
+
+## Defining a Repository Class
+
+A repository object gets access to selected relations that are available in the
+ROM container object. When you create a repository class you can explicitly define
+which relations should be established:
+
+``` ruby
+class UserRepository < ROM::Repository::Base
+  relations :users
+
+  # your methods go here
+end
+
+rom = ROM.finalize.env
+
+user_repo = UserRepository.new(rom)
+
+user_repo.users
+```
+
+Once you have access to your relations you can expose access to specific entities
+with your custom interface:
+
+``` ruby
+class UserRepository < ROM::Repository::Base
+  relations :users
+
+  def [](id)
+    users.where(id: id).one!
+  end
+end
+
+
+rom = ROM.finalize.env
+
+user_repo = UserRepository.new(rom)
+
+user_repo[1]
+# #<ROM::Struct[User] id=1 name="Jane">
+```
+
+As you can see the repository automatically maps relation tuples to instances of
+`ROM::Struct`.
+
+## ROM Structs
+
+ROM structs are simple data-capsules. They give you access to their values via
+attribute `[]` readers and are coercible to a hash. With powerful relation query
+DSLs you can easily compose relations and return data in the expected state.
+That's why automatically mapping to small structs is feasible and no custom
+mapping is required.
+
+Struct constructor is strict with regards to the attribute hash it receives - if
+there are attributes that were not expected you will get an exception.
+
+Here's an example of a struct:
+
+``` ruby
+user = user_repo[1]
+# #<ROM::Struct[User] id=1 name="Jane">
+
+user.id # 1
+user[:id] # 1
+
+user.to_hash # {:id=>1, :name=>"Jane"}
+```
+
+ROM repositories generate struct classes for you so that you don't have to define
+them.
+
+<div class="well well-small">
+There's a planned feature where you'll be able to provide namespace for your
+structs where struct class constants will be defined as opposed to anonymous classes.
+</div>
+
+## Working With Relations
+
+In ROM relations represent specific database queries that your application needs
```

Comment:
In ROM, relations represent specific database queries that your application needs
to execute to get its data. It is recommended that you define explicit relation
classes to achieve better encapsulation and simplify composition.

A common pattern is to encapsulate specific relation methods and use those in your
repositories rather than accessing lower-level query DSLs.